File: typescript/Functional Programming/SumTypes.ts
(() => {
    type LinkedList = Node | Empty;

    class Node {
        kind: "Node" = "Node";
        constructor(public value: number, public next: LinkedList) {}
    }
    
    class Empty {
        kind: "Empty" = "Empty";
    }
    
    function processList(list: LinkedList): void {
        switch (list.kind) {
            case "Node":
                console.log(`Node with value: ${list.value}`);
                processList(list.next);
                break;
            case "Empty":
                console.log("End of list");
                break;
        }
    }
    
    // Example usage
    const myList: LinkedList = new Node(1, new Node(2, new Node(3, new Empty())));
    processList(myList);
    
})()

File: python/Functional Programming/SumTypes.py
from dataclasses import dataclass

@dataclass
class Node:
    value: int
    next: 'LinkedList'

@dataclass
class Empty:
    pass

# Using the union operator to define the LinkedList type
LinkedList = Node | Empty

def process_list(lst: LinkedList):
    match lst:
        case Node(value, next):
            print(f"Node with value: {value}")
            process_list(next)
        case Empty():
            print("End of list")

my_list = Node(1, Node(2, Node(3, Empty())))
process_list(my_list)
